Gsdx black fmv

gregory38 opened the pull request:
Commits fix rules of roses and I hope various black screen FMV

Performance impact must be tested, and I'm afraid of strange texture cache behavior.

ssakash commented:
Your [third commit](https://github.com/PCSX2/pcsx2/commit/256276912a82ebca49fd259d120be6c30d1ef0b0) causes some graphical issues on Bomber man Kart and also random crashes at few instances.

**Compiled without third commit:**
![fine](http://ultraimg.com/images/gsdx_20160306053022.png)

**compiled with third commit:**
![not so fine](http://ultraimg.com/images/gsdx_20160306053137.png)

ssakash commented:
[Third commit](https://github.com/PCSX2/pcsx2/commit/256276912a82ebca49fd259d120be6c30d1ef0b0) also causes random crashes on _**Ben 10 : Protector of earth**_. though without any graphical glitches. The issue isn't reproducible with the first two commits.

gregory38 commented:
Only the latest commit impact the rendering.

I updated the code to support 16 bits format. And to check the size of the upload. I'm not sure it will fix the crashes but it won't be worst.

ssakash commented:
the new commit seems to fix to all the regressions, I'll try doing an indepth test later.

gregory38 commented:
Oh great. Today is my ludky day. Texture blocks are 2 time smaller in 16 bits, so I guess it was overflown.

By the way, it must be tested without the "Switch to GSdx SW when an FMV plays" gamefix. Hopefully it will become useless.

MrCK1 commented:
I'll test GT4 later; just hope it will get past the copyright screen :(

refractionpcsx2 commented:
I doubt this will fix GT4, that doesn't even work in software mode, the video doesn't even actually play as far as I'm aware.

ssakash commented:
> I doubt this will fix GT4, that doesn't even work in software mode

I assume it's a core issue ?

FlatOutPS2 commented:
This commit hasn't fixed any software mode-only video's I've tried. I still needed to f9 to software mode to see the video's.

bositman commented:
@FlatOutPS2 Please explicitly state games so we don't test the same games again...

gregory38 commented:
@FlatOutPS2 
Please send me a gs dump for a game. I put very strict condition on the texture cache. Maybe I need to relax them a bit. Or maybe I miss somethings in the texture cache.

FlatOutPS2 commented:
Here are two video GS dumps that only play in software mode:
http://forums.pcsx2.net/attachment.php?aid=59318

gregory38 commented:
Thanks you. I can reproduce the bug on my side. Now remains to understand the issue and to fix it.

gregory38 commented:
Hum, trailer seems to work with the preload hack (at least on the dump)

gregory38 commented:
On the others, it is very interesting. @ssakash will love it.

So the game uses an internal resolution of 640x512, very standard. But it draws the texture on the rectangle (0 , 992 ) x (640, 1232). In native resolution the FB is 1280x1024... So you can understand that texture is drawn in the void....

Currently the source texture is located at 0x1400 and the FB at 0x0 (FBW == 10). The pixel (0) x (992) is therefore located at (992/32) == 31th height page. A page row is 320 blocks. So we are around 0x26C0 which mean in the source texture. In others word, the game overflow the frame buffer to draw into the texture. It copies (blits) data from the top of the texture to a lower part...

Now I can search a rope to hang myself :p I dunno how to fix it. Let's try a bigger FB to see if it "fixes" the bug. Good new the bottom part is correctly drawn in the FB. Bad new the texture isn't invalidated so there is a texture cache bug too.

gregory38 commented:
Ok I understand the texture cache bug.
What happen after the draw
1/ invalidate the texture
2/ upload the data from the GS memory (CPU) to GPU texture

What ought to be the correct version
1/ invalidate the texture
2/ transfer the rendering from the VRAM to the RAM (aka GS memory)
3/ upload the data from the GS memory to GPU texture

gregory38 commented:
FWIW, I added a check on FB size. And it seems that the "trailer.gs" also draw outside of the FB, namely between 1024th line and 1280th line. I give it a closer look and it seems a similar pattern as above.

I really don't know how to fix it. So far, it seems FMV rendering is done after the line 1024, so maybe I can try to detect the pattern and completely invalidate the RT target. It won't be fast, but it could be enough for a couple of FMV. Boys, it is time to find game that suffers of this issue, so I can polish my idea.

gregory38 commented:
I pushed a tentative fix for native resolution. Please test some games @ramapcsx2, @FlatOutPS2

FlatOutPS2 commented:
It doesn't fix it. It did show some vertical lines dancing on the screen at times, but it's stil black most of the time.

gregory38 commented:
Just to sure, you're in native mode.

FlatOutPS2 commented:
Yes, it's completely black at higher resolutions.

gregory38 commented:
Ok. The issue is the texture cache management. I suspect I fix the bug in the middle of the flow, it means that if the texture cache is correct at instant T of the video, the next frame will be correct. However I suspect a bug remains in the start of the video. First frame is likely wrong, and it is likely reused. I would need a bigger gs dump, one that start before the FMV so cache is polluted with wrong data.

FlatOutPS2 commented:
Is this enough?
http://forums.pcsx2.net/attachment.php?aid=59329

gregory38 commented:
Did you try with the preload hack?

FlatOutPS2 commented:
Yes, same result.

gregory38 commented:
Hum I slowed down the rendering, and I spot a couple of glitches. However I'm far from a black screen.
![screenshot - 03122016 - 10 23 59 am](https://cloud.githubusercontent.com/assets/1569559/13722182/99d485d2-e83c-11e5-9bed-e8808ae78b74.png)

Could you tell me the complete name of the game? I will try to find it in my local resellers.

gregory38 commented:
I have a trail. Here a texture cache lookup of the frame.
```
Message:TC: Lookup Frame 1280x1280, close hit: 21 (0x800, took 0x0 -> 0x2800)
Message:TC: Lookup Frame 1280x1280, perfect hit: 1 (0x800 -> 0x1000)
```
We try to load data from pointer 0x800. I can you see there are 2 valid buffers. One that ranges from 0  to 0x2800 and the others one that ranges from 0x800 -> 0x1000. The texture cache take the 2nd one. But I think it mustn't exists, because it is a subpart of the first buffer. Indeed if I check the texture, the small buffer is fully black.

gregory38 commented:
The issue appears at the start of the video. And I think I potentially found why there is garbage instead of nice black bar on some video.

In this example the game draws the menu at 0x800 to 0x1000. Before the video, it will do a clear of the zone. For GSdx it is the RT1.

Then it will use the RT located at 0x0 to 0x2800 to draw the FMV. However GSdx sees the new RT as a separate buffer.

So in memory, you have this pattern. 
```
RT1: _________000000000_________
RT2: VVVVVVVVxxxxxxxxxxVVVVVVVV
```
`_` is don't care, `x` is random data, `0` is dark pixel, `V` is valid data.

Whereas it ought to be
```
RT1: VVVVVVVV0000000000VVVVVVVV
```
So great we have 2 bugs :(

gregory38 commented:
I pushed new commit. Please tell me that it is working now :)

FlatOutPS2 commented:
Yes, it's now working perfectly on native resolution.

gregory38 commented:
Cool :smiley: 
Now, I would need a similar dump of a game that suffers of garbage in the top/bottom bar.

gregory38 commented:
@bositman @refractionpcsx2 @ssakash, all I need your opinion

Games that suffer of black FMV requires a bigger framebuffer, i.e a height of 1280. We could set the value as minimum but it will increase the memory requirements on others games such as Gran Turismo. Or likely any game that uses an offset sub framebuffer. Instead to reuse the framebuffer with an offset, GSdx will create hundreds of new framebuffers. How do we solve this mess?

Some random ideas
1/ goes 1280 for 2x (and maybe 3x only). Bigger upscaling will use standard height
2/ goes 1280 for all factors. And try to find solution to skip bad draw call that requires an useless framebuffer.
3/ Add a gsdx option to replace the one from the core.

gregory38 commented:
4/ detect the write above 1024 and extend the size of the target.

ssakash commented:
> 4/ detect the write above 1024 and extend the size of the target.

I vote for this method, detecting such cases and extending the size when necessary seems much better.

willkuer commented:
Gsdx option maybe as crc hack could also be reasonable.

gregory38 commented:
I added a dynamic detection. I was afraid of the complexity but actually it was easy. @FlatOutPS2 could you test again your games with various upscaled resolution. 

Unfortunately, I think I have a regression on the trailer.gs.

chane2k1 commented:
Just woke up, here's the GS dump for the Kingdom Hearts 1 fmv garbage, https://drive.google.com/file/d/0B-XoRQUIxD4tS1hla0FnM0JuTHc/view?usp=sharing

gregory38 commented:
Thanks. I will give it a look.

FlatOutPS2 commented:
The video's play just fine when upscaled, even when switching resolutions while the video is playing. But when I went to the Video Plugin settings menu during the Spy Hunter and Trailer video's and cancelled/escaped out of the menu, the video starts flickering. This doesn't happen when you do change resolution or even just press OK without changing settings. This only happens with the OpenGL hardware renderer for the Formula One 2001 video.

At even resolutions(2x,4x,6x,8x) in OpenGL Hardware mode a small constant vertical stripe appeared at the right of the trailer video, it isn't there at uneven resolutions(3x,5x). 
A complete strip from top to bottom shown at times at the right side of the Spy Hunter video's too(at all scaled resolutions). It looks like it's the first line of the left side of the video is shown as the last line on right.  A similar effect appears with the Formula One 2001 video, but it also has a smaller constant horizontal stripe at the bottom at all scaled resolutions, which also appears with DX11, unlike the other stripe.

At all scaled resolutions in both DX11 and OGL video's arent as smooth as on native/software mode. A lot of hard edges appear.

gregory38 commented:
hum, ok for the gui but I won't fix it. The texture cache isn't coherent with the real data to keep speed acceptable.

I need to check, but Formula 1 put the depth buffer at the same location as the texture. There are likely some bad interactions

```
At all scaled resolutions in both DX11 and OGL video's arent as smooth as on native/software mode. A lot of hard edges appear.
```
The game write the texture in native. Then we draw the texture, so result is upscaled. Then I downscale the draw to copy back on the GS memory.Which then reupload as native to be drawn at the final destination (upscaled). The GPU->CPU transfer is barely used. So potentially there are bugs on it on openGL.

Note: it seems the game uses 2 blits to rescale the image.
If we want the best quality, we need to avoid all those rescalings. Maybe we can shortcut the draw, and draw directly the texture into the final destination. Let's me check it.

gregory38 commented:
Let's do a summary, games have this pattern:
````
-------------------------
| Frame 1 (512x256)    |
-------------------------
| Frame 2 (512x256)    |
-------------------------
| Texture (512x512)    |
|                      |
-------------------------
| Bottom T (512x256)   |
-------------------------
```
Note: GSdx have a single Texture for both "Texture and its bottom". And a single target for the 4 zones.

Odd frame:
* EE writes FMV into `Texture`
* `Texture` is blitted into `Bottom T`
* `Bottom T` is blitted into `Frame 2`

Even frame:
* `Texture` is blitted into `Bottom T`
* `Bottom T` is blitted into `Frame 1`

The idea solution will be to upload directly the FMV to the back buffer. It would avoid the GS resizing, and the increase of the RT buffer.

Another solution will be to do the copying manually from `Texture` to both `Frame 1` and `Frame 2`. And let's the standard merge output operate. Less hackish and potentially easier to implement.

gregory38 commented:
I implemented the 2nd idea as a hack which replace standard CRC hack. If it works well, we will improve it (CRC/option/always on). Could you give it a shot?

FlatOutPS2 commented:
The previous flickering issues remain, and some of the video's are now moved to the left creating a black bar on the right and they're constantly flickering in hardware modes without going to go to the menu or pressing f9 first.

Hard edges have improved, the difference between SW and HW video's isn't as pronounced as it was before.

gregory38 commented:
```
moved to the left
```
Oh, actually I upload the texture at first pixel. I guess I need to shift it.

However no idea for the flickering :(

Let's try something different. Instead to move in the final destination. I will move to the bottom of the texure.

Wait. I removed the invalidation of the sub rt. Maybe I just need to put it back. Give me 30 seconds.

gregory38 commented:
Done. I need to think more about the left stuff.

gregory38 commented:
Actually it won't work. I need to do it differently.

gregory38 commented:
I think I'm mostly good now. If you still have issue with left aligned FMV, please post me a small gs dump.

For the flickering, it won't be perfect. I think it could miss the first frame of the video, if you interrupt it.

FlatOutPS2 commented:
The left alignment error is gone, but the flickering after going to the plugin menu/pressing F9 is still present.
But the image is completely black again in all DX11 Hardware resolutions(including native), with only the Formula One 2001 video showing some garbage mostly near the top.

gregory38 commented:
Damned !

gregory38 commented:
Hum, could you quickly test if it is better with the above commit. Normally formula 1 must be fine.
If you still have a black video on others game, try to generate me a long dump (like you did for F1). I think I understand where the regression happen. It would be easier if I can test in local.

FlatOutPS2 commented:
It didn't change anything. Formula One is still black with the static garbage too.

gregory38 commented:
Ok. Thanks you for your time. It is very annoying that dump are working on my side. I think my mistake was to bypass the bad draw. But I still need it to properly handle the invalidation.

gregory38 commented:
So take 10. I hope it will work now, or I will lose my sanity ;)

Anyway, could you test first opengl and native. (then upscaled and finally DX)

FlatOutPS2 commented:
Prepare to lose your sanity, cause I still get the same results.

gregory38 commented:
Seriously ! Can you list your tested game ? I need to find one on the market.

FlatOutPS2 commented:
Do you have any demo/preview discs? Because many of those have trailers with the same issues.
The games include:
Spy Hunter
Formula One 2001
Kengo

gregory38 commented:
PAL or NTSC?

FlatOutPS2 commented:
All PAL

gregory38 commented:
Hum, are demo disk shareable from a copyright PoV? Here a screenshot of Kengo (opengl native) with the branch. Master is completely broken, but it works with the branch. The only minor issue is a line on the right. Strange. Are you sure to have latest build and the good config?

![kengo](https://cloud.githubusercontent.com/assets/1569559/13796743/b06c7182-eb08-11e5-80be-3430e2fb5ce6.png)

prafullpcsx2 commented:
Why you ain't using block dumps?

FlatOutPS2 commented:
@gregory38 As I've mentioned, the black screen issue is with DX11, not OpenGL.

gregory38 commented:
@prafullpcsx2 yes. It was a rhetorical question :) By the way, check your regression thread.

@FlatOutPS2 
I misunderstood you. So potentially I don't need the invalidation stuff.

However the DX issue is likely in the read back stuff.Is MSAA enabled on DX ? I guess DX doesn't support write to a texture....

refractionpcsx2 commented:
> I guess DX doesn't support write to a texture....

it does as far as I know.

gregory38 commented:
Dunno GSdx has texture/render target and offscreen. Modern OpenGL has only textures. (used to have render target)

refractionpcsx2 commented:
http://www.gamedev.net/topic/263289-editing-a-dynamic-directx-texture/

gregory38 commented:
Well GL is fine but not DX ;)

Hum, could be that
```
		m_dev->StretchRect(tex->m_texture, sRect, tex->m_texture, dRect);
```
Read/Write to the same texture. Maybe I need to copy it first.

willkuer commented:
Is preload data frame active for dx as well? I guess you are testing both with preload data frame in ogl? This would be at least a simple solution.

gregory38 commented:
@willkuer it must work. But I'm sure it is a dx limitation. I just need to copy data around. Anyway a purely SW approach will be better.

@gabest11 
Hello. I need your precious help on 2 topics
1/ Is there a way to copy odd or even lines from a rectangle GS mem to a packed rectangle? Otherwise I guess, I need to update the m_valid of the texture to avoid the 2nd upload.
Use case: this issue. As a summary, GS blit from one part of the texture to another through a draw.
2/ Is there a way to copy a constant value into a rectangle? (Typically a 0)
Use case: garbage in top/bottom bar of FMV. Clear is done in the RT but not the GS mem, which then got reuploaded after a EE texture write.

gregory38 commented:
I tryied to do something for dx11. Tell me if it is working.  (it is only a stab in the dark)

FlatOutPS2 commented:
Again no change.

gregory38 commented:
Sigh... I won't be able to fix the DX code. @refractionpcsx2 do you want to give it a shot?

refractionpcsx2 commented:
What makes you think *I* can fix the DX code? last time I was just copying you :P

gregory38 commented:
Because I believe in you :) (and because you have access to a GPU DX debugger).

First step is to check the content of the texture after the blit
```
m_dev->StretchRect(tex->m_texture, sRect, tex->m_texture, dRect);
```
Basically, the bottom of tex->m_texture must be a copy of the top. FlatOut have posted some gs dump (far above)

refractionpcsx2 commented:
I may look at it at some point but I have little knowledge of dx and haven't a clue how to use the vs dx debugger lol

gregory38 commented:
Ok. So I will move a bit the code around. So we will have a dedicated read function per renderer. And maybe check if I need the extra invalidation. I really want to avoid it.

gregory38 commented:
I massively clean the code. I change the code to be more DX friendly.
Remaining to do list
* DX11 behavior.
* Check trailer.gs potential regression
* Usefulness of the texture cache invalidation
* Properly enable the code. Option/CRC/Always?

FlatOutPS2 commented:
Video's now work in DX hardware modes again. :)

There's still the small issue with lines and the more serious flickering and a darker image that often occur when f9'ing or exiting the video plugin settings.

refractionpcsx2 commented:
Good job Gregory :) See you don't need me :P

gregory38 commented:
See when I said the API is completely dull and can't draw into a texture.

Opengl
```
 m_dev->StretchRect(tex->m_texture, sRect, tex->m_texture, dRect);
```
Versus DX
```	
		GSVector4i r_full(0, 0, tw, th);
		if (GSTexture* rt = m_dev->CreateRenderTarget(tw, th, false)) {
			m_dev->CopyRect(tex->m_texture, rt, r_full);

			m_dev->StretchRect(tex->m_texture, sRect, rt, dRect);

			m_dev->CopyRect(rt, tex->m_texture, r_full);

			m_dev->Recycle(rt);
		}
```

```
There's still the small issue with lines and the more serious flickering and a darker image that often occur when f9'ing or exiting the video plugin settings.
```
You're asking too much :p Which line? Honestly F9 is a lower priority. Did you try with the preload hack (in case it have any impact).

Otherwise I have a fix for kingdom heart coming very soon.

gregory38 commented:
@refractionpcsx2 could you test kingdom heart? Add in the ini file the new hack
```
UserHacks_GsMemClear = 1
```
I don't know what is the best solution. An option/a CRC/always enable it but in this case it will worth a couple of benchmarks.

refractionpcsx2 commented:
Comiling now, will let you know shortly

refractionpcsx2 commented:
hmm, i cant seem to compile this branch, complaining that windows.h is missing.

refractionpcsx2 commented:
ok do i need to enable anything else? cos i still get junk top and bottom.  I enabled hw hacks but made sure nothing was enabled in there then added UserHacks_GsMemClear = 1 to my gsdx.ini

gregory38 commented:
(I'm so glad that I'm on linux ; )

Ah great :) One less bug. Any opinion for the option/crc/always on?

refractionpcsx2 commented:
That depends, if it's not going to cause problems anywhere else (or if it will help somewhere else) then I would just say make it always on.

Can you think of any other scenarios/games which would potentially hit it that can be tested?

gregory38 commented:
The code intercept the rendering of a black sprite. And instead to only draw the sprite in the GPU buffer. It also memclear the associated gs memory. The only impact is potentially the perf. I said potentially because it is likely done only a couple of time by frame.
Bosit said severals game suffer of those fmv issues. But I don't have any details.

refractionpcsx2 commented:
well i would personally have it permanently on, with maybe a hidden ini entry to disable it if somebody comes across an issue?

gregory38 commented:
I like you idea :)

Let's go for a reverse hack.

gregory38 commented:
Done. Remains to test and merge the code.

FlatOutPS2 commented:
> You're asking too much :p Which line? Honestly F9 is a lower priority. Did you try with the preload hack (in case it have any impact).

It's not just f9, it can happen every time you go change some GSdx setting, like turning on the preload hack(or even just cancel out of it). The preload hack seems to make it worse, adding a ghosting effect to the flickering. Really, it's seizure inducingly bad.

gregory38 commented:
Geez, you saw the half empty bottle :) It is already a big progress to be able to see the fmv in one shot.

Maybe I have some ideas of what could be wrong. But ot is hard to be sure and hard to fix.

bositman commented:
@FlatOutPS2  I still see flickering hell in Dynasty Warriors 3 (E), even in native and OpenGL renderer, any special ini setting im supposed to be using?

FlatOutPS2 commented:
I retested and got the flickering again, so something must've gone wrong(I did notice rebooting the game doesn't always reintroduce the bug, you need to quit PCSX2 before it returns). I didn't use any special settings as far as I know.

bositman commented:
Yeah any kind of pause/ F9 toggle will fix it until next time you open PCSX2, so I guess that's unsolved yet.
About Kingdom Hearts, I'm not getting the bug in master or in this PR :/

alfio8788 commented:
Since pause/F9 seems to fix this, can't it be simulated automatically as temporary fix/hack till a better fix/workaround is found?

bositman commented:
We already have a game fix that auto switches to SW rendering when an FMV plays.

gregory38 commented:
Yeah. Goal is to fix hack.

Anyway, I'm currently playing with the frame lookup of the texture cache. If the texture cache picks the wrong buffer, it will draw a black screen. I think it could explain the F9 issue on the above case. And maybe DW3

gregory38 commented:
```
Type:Others	ID:6	Severity:Info	Message:TC: Lookup Frame 1280x1024, miss (0xa00)

Type:Error	ID:20	Severity:Mid	Message:TC: Lookup Frame 1280x1024, inclusive hit: 1 (0xa00, took 0x0 -> 0x1360)
```
Here an example from the spyhunter gs dump. On the target/frame 0xa00. As you can see on the draw number 6. We have a cache miss. However few draw later, on the draw 20, the content of the frame 0xa00 is actually on the frame 0 (but with an offset).

The potential issue is that miss will create a "wrong" 0xa00 frame which could later be reused.

gregory38 commented:
@ALL Could retry the black FMV test flickering.

refractionpcsx2 commented:
There's gonna be some poor soul called "All" on Github now who just got pinged :P

gregory38 commented:
:)

@refractionpcsx2 
By the way, I added a speed hack for snowblind engine game. In case, you misses it, you could enable it with `UserHacks_DisablePartialInvalidation = 1`.

refractionpcsx2 commented:
Cool thanks :) I thought I'd already put it in my ini, apparently I have not, I did wonder why memory usage seemed exactly the same (never went over 1gb for me btw, but it was slowly creeping)

gregory38 commented:
Hum. Actually I don't know what is the impact of the hack on the memory. Potentially it could reduce it.

Initial purpose was to reduce the book-keeping overhead. But it would delete the texture that are repeating so potentially the big one. I could explain the massive speed up with 32 bits textures.

gregory38 commented:
Indeed you're more than right. On my Baldur Gate GS dump. Here the stat in Native.

8 bits
```
Message:MEM: RO Tex 159MB. RW Tex 0MB. Target 75MB. Depth 10MB
Message:MEM: Surface Pool 88MB
```
32 bits
```
Message:MEM: RO Tex 1357MB. RW Tex 0MB. Target 75MB. Depth 10MB
Message:MEM: Surface Pool 477MB
```

8 bits + hack
```
Message:MEM: RO Tex 1MB. RW Tex 0MB. Target 75MB. Depth 10MB
Message:MEM: Surface Pool 25MB
```
32 bits + hack
```
Message:MEM: RO Tex 8MB. RW Tex 0MB. Target 75MB. Depth 10MB
Message:MEM: Surface Pool 49MB
```

I think, I don't need to explain the number :p The hack manage to save 1.5GB of texture in my 38 frames gs dump :grinning:

FlatOutPS2 commented:
The flickering seems to be gone, but a static ghost image appears over the video instead. And it introduces a regression in many games, creating vertical and horizontal lines on 2D screens, and next to letters in text messages.

gregory38 commented:
hum maybe it wasn't a good idea to force the preload hack for the frame. Could you edit GSTextureCache.cpp line 504 and remove the `true || ` part.

FlatOutPS2 commented:
That removes the ghost image over the video. :smile: Doesn't affect the regression.

refractionpcsx2 commented:
> I think, I don't need to explain the number :p The hack manage to save 1.7GB of texture in my 38 frames gs dump  (for 32 bits textures)

indeeed you don't! That's an impressive fix! Hopefully that is the bulk of the memleaking and should improve the slow areas! :)

What a fruitful PR ^_^

gregory38 commented:
@refractionpcsx2 
Actually the snowblind hack is already in master ;) And it isn't a memleak. It is due to the strange GS texture format.

@FlatOutPS2 
Ok :( I don't understand why it break the rendering. Could you provide a gs dump, likely a long dump. Or potentially there is another bug.

FlatOutPS2 commented:
Nevermind, it's not a regression(at least, not because of this commit). 
The issue only appears with the default GSDX PCSX2 starts with without changing settings(before creating a GSDX.ini, so D3D11 HW on native). As soon as you go into the the video - plugin settings and select one of the hardware renderers, even D3D11 or 9 HW on native and press OK, the issue is gone. I deleted the GSDX.ini in master and stable 1.4.0 and the same thing happens.

gregory38 commented:
I'm feeling much better :) By the way, what is the status on Dynasty Warrior, still broken I guess ?

FlatOutPS2 commented:
Actually, it looks like it really is fixed this time, no issues with a fresh PCSX2.

gregory38 commented:
woot so much improvement this week !

refractionpcsx2 commented:
> woot so much improvement this week !

You're on a mission! xD

gregory38 commented:
```
You're on a mission! xD
```
Oh yes, I improved the snowblind engine hack. Another 2x speed boost on Champion of Norrath.

bositman commented:
Hm Dynasty Warriors games have stopped flickering but there is a new issue now, it looks like it's missing frames or bad interlace, I can't explain it easily but it is definitively very visible

gregory38 commented:
You're killing me. Could you give me a gs dump.
And yes the game likely lookups the wrong frame.

bositman commented:
Sure, single frame or multiples?

bositman commented:
It's a couple of frames, done in OpenGL SW using the plugin from master:
http://bositman.pcsx2.net/dumps/dynasty_warriors3_pal_FMV.rar

prafullpcsx2 commented:
I am not 100% sure but these changes seems to have fixed https://github.com/PCSX2/pcsx2/issues/740

gregory38 commented:
Ok

FlatOutPS2 commented:
The latest commit causes a crash of PCSX2 during ingame cutscenes in Ace Combat 04. The VS debugger mentions GSTextureCache.cpp InvalidateVideoMem.

gregory38 commented:
@bositman 
I don't see the issue. Do you have any hack enabled? I have the game, I will try to reproduce it.

@FlatOutPS2 
Do you have more info, like the line number? Or a backtrace.

bositman commented:
Strange. No hacks on, happens with scaling and native. GSdx is on OpenGL HW, hardware depth, high accurate blending, accurate date on. Everything else is on default.
It happens in the intro video after you select language so it's easy to reproduce if you got the game

gregory38 commented:
My version is the Dynasty Warrior 3 - Xtreme Legends (PAL FR).
```
I can't explain it easily but it is definitively very visible
```
I'm afraid  you will need to explain it to me. Or post a recording/screenshot.

The only thing, I see is the low quality (blocky) of the video.

FlatOutPS2 commented:
AC4 crashes at: GSdx32-SSE2.dll!GSTextureCache::InvalidateVideoMem(GSOffset * off, const GSVector4i & rect, bool target) Line 764

Best way to reproduce the bug is skipping a in-game cutscene midway.

gregory38 commented:
```
AC4 crashes at: GSdx32-SSE2.dll!GSTextureCache::InvalidateVideoMem(GSOffset * off, const GSVector4i & rect, bool target) Line 764
```
Could you print the line?

FlatOutPS2 commented:
You mean this?:
int y = GSLocalMemory::m_psm[psm].pgs.y * offset / rowsize;

gregory38 commented:
Ah ah. Maybe rowsize is 0

gregory38 commented:
@FlatOutPS2  Check the last commit

FlatOutPS2 commented:
That solved it. :smiley:

gregory38 commented:
great. Remain to wait bosit validation of the half fps bug. And check regression of latest commit

refractionpcsx2 commented:
I don't know if its relevant as you're addressing offsets on the video, but the Twisted Metal (not Black, the other one) videos are off to the left, hope this will help http://www.filedropper.com/fmv-ofst-twistedmetal

bositman commented:
Sorry, half FPS and flickering is back :P

gregory38 commented:
We're making progress

gregory38 commented:
The flickering is due to a self-merge of the frame with itself. I just need to find how to fix the condition (in a way that work for both renderer)

FlatOutPS2 commented:
@refractionpcsx2 That should be fixed too, as I encountered a similar cut off on the right video that wasn't cut off anymore with one of the latest commits(before the current regressions).

gregory38 commented:
New commit for @bositman

gregory38 commented:
@refractionpcsx2 did you generate the dump on the SW renderer?

refractionpcsx2 commented:
I didn't,  I must admit, but the result was the same

gregory38 commented:
SW is broken too ?

refractionpcsx2 commented:
Yup

gregory38 commented:
@ssakash could you please check the dump of ref above.

bositman commented:
Dynasty Warriors 6 flickering gone, half fps remain :P

gregory38 commented:
Are you sure? Sigh. I need to regenerate a full trace but I'm nearly sure that behavior is better

gregory38 commented:
Few typos later :) I hope it is good now.

bositman commented:
Hey you made it! XD All Dynasty Warriors games work fine even with scaling, also tested a couple of other games with videos and they were fine. Good job

gregory38 commented:
:wine_glass:

refractionpcsx2 commented:
Good job! :+1:

willkuer commented:
It sounds like you resolved all remaining gs issues with one pr.


Well... maybe not all but still a nice progress here!

bositman commented:
Since you're doing so well, here is another strange FMV issue of HW renderers :P
Shadow Hearts 2, videos work fine but flicker only when someone speaks 0_o 
Bug not present with SW renderers

Video of issue: https://www.youtube.com/watch?v=lbOLYkitlQA
Dump: http://bositman.pcsx2.net/dumps/shadow_hearts_2_flicker_FMV.rar

gregory38 commented:
```
It sounds like you resolved all remaining gs issues with one pr.
```
GS is the modern version of the `Danaïds' sieve` ;)

```
Shadow Hearts 2, videos work fine but flicker only when someone speaks 0_o
```
My guess is the blending of subtitle.

```
Dump is a bit huge since I wanted to grab frames before the flicker and after the flicker to help you more :P
```
You're a SSD killer. Could you post also a single frame dump during the flicker too ? It is very slow to replay full gl dump and log are huge. So if I can reproduce the issue with a small dump, it will be nice :)

bositman commented:
Sure, I'll provide one when I get home in a couple of hours ;)

gregory38 commented:
take your time. I'm not home neither, and I won't have any free time soon.

FlatOutPS2 commented:
Kengo: Fine at native flickering appears only after using f9 a lot, flickering appears quickly at higher than native resolutions. 
trailer video's: Same as Kengo.
Spy Hunter: Same as above.
Formula One 2001: Same as above, but at 2x native 2 duplicate smaller frames of the video are spammed over the left side of the screen.
DW3: Flickering and vertical lines garbage over the video's
AC4: Doesn't crash anymore.

And a new case:
X-Files Resist or Serve: Right side of some video's are cut off and on flickering(this seemed fixed in earlier commit). The intro video when starting the first episode still flickers.

bositman commented:
Huh? I tested Dynasty Warriors 3 and it was perfect, recheck settings and/or if you're using the latest commit

FlatOutPS2 commented:
Definitely the latest commit. I rechecked and it seems to only affect higher than native scalings.

bositman commented:
Bleh you are right. Dynasty Warriors 4,5 and 6 work perfectly even with scaling, Dynasty Warriors 3 only in native, scaling makes it go wacko.

gregory38 commented:
Well I only tested native. Maybe I misses a scaling factor somewhere. Or worse, I need to increase the size of the rendering buffer. But it will greatly increase the memory consumption on a couple of games. Or maybe we could fix those games too.

bositman commented:
Here is the small dump for shadow hearts 2: http://bositman.pcsx2.net/dumps/shadow_hearts2_flickering_fmv_small.rar

gregory38 commented:
For the upscaling issue, could someone test a quick change?
Edit: GSRendererHW.cpp (around line 61)
Replace
```
	int fb_height = (fb_width < 1024) ? max(512, crtc_size.y) : 1024;
```
by
```
	int fb_height = 1280;
```

FlatOutPS2 commented:
The change fixes the DW3 scaling issue and the X-Files cut-off video's seem corrected as well, though not the episode video flickering issue.

gregory38 commented:
@bositman did you try the preload frame hack?

gregory38 commented:
No need to test above commit. It doesn't work as expected...

gregory38 commented:
Honestly, I'm lucky if this PR does the job correctly. Anyway, I increased the size of the framebuffer (* 2). It will be killer on some games but we don't have choice here. I'm afraid we will need to tune game that requires hundreds of frame buffer to avoid the huge memory requirements.

turtleli commented on plugins/GSdx/GSRendererHW.cpp:
```diff
@@ -58,7 +58,11 @@ void GSRendererHW::SetScaling()
+	// Gregory: the framebuffer must have enough room to draw
+	// * at least 2 frames such as FMV (see OI_BlitFMV)
+	// * high resolution game such as snowblind engine game
+	int fb_height = max(1024, 2*crtc_size.y)
```
Missing `;`)

ssakash replied:
> int fb_height = max(1024, 2*crtc_size.y)

no other alternatives to dynamically increase the size when necessary ? :/

gregory38 replied:
@turtleli 
Please don't follow my example ;)

@ssakash 
I tried to do something. It is far from ideal, it will lose a couple of frames after a pause. It would be so much better to fix the texture cache allocation crazyness. And then drop all this mess.

gregory38 commented:
Dear Testers,
Could you make a summary of the current state of the PR?

FlatOutPS2 commented:
There's still flickering after going to the video plugin settings or pressing f9, occuring more often than it did a few commits ago, especially at higher than native resolutions. Seems less frequent with OpenGL.

Kengo: Fine, garbage at the first frame(s).
Trailers & Spy Hunter: Fine
Formula One 2001: Garbage at the first frame(s). When the flickering occurs at 2x native, 2 duplicate smaller frames are spammed over the left side of the screen.
DW3: Fine

New case: 
James Bond 007: Everything Or Nothing: Video's have constant flickering with, at higher than native resolutions, horizontal lines/gaps in the image that grow per scaling setting. Occurs in PR and stable 1.4.0. Similar to the other video's it's fine in software mode.

gregory38 commented:
Thanks for your report. I will revert my change for the auto-detection of the height. It requires 2 frames to stabilize, and just fill the texture cache with junk.

gregory38 commented:
I reverted my auto detection code. 

On 007, I don't have the flickering on the dump. And the gap is normal. The game renders the FMV as line. FF DoC uses points. Hack is possible (by replacement of the line by some sprite with the CPU or a Geometry Shader)

Aced14 commented:
Here's another game that's still suffering from FMV flickering (as of 8943a9db4a43c1f3ec78ca176eb42bf9a5c5e68a)...

**Monster Attack**
  * [GS dump (OpenGL hardware)](https://Aced14.github.io/pcsx2-gs-dumps/ma_gs_fmv-flicker-bug.7z)
  * NTSC-J version of this and its sequel have the same issue in all their FMVs.

gregory38 commented:
@Aced14 I think your dump is too small. I can't reproduce the issue on my side.

Aced14 commented:
@gregory38 Here's a multi-frame GS dump of Monster Attack:
[Monster Attack: Multi-frame GS GS dump (OpenGL hardware)](https://Aced14.github.io/pcsx2-gs-dumps/ma_gs_fmv-flicker-bug-2.7z)

Just to provide more information, when this game starts up, it checks for memory card data, then proceeds to show company logos. The first logo is a static image for "agetec" (fades in/out and works fine in all GSdx modes and renderers). The second logo is for "D3 PUBLISHER", which uses an FMV. In GSdx hardware (both OGL and DX11), that particular FMV flickers heavily between black and its "real" video frames. It doesn't flicker in GSdx's software modes.

gregory38 commented:
@Aced14 could you provide me a short iso dump? Try to search on forum how to generate it, I don't remember which plugin can create it. I didn't manage to reproduce the issue with the GS dump. 

Otherwise I think I will merge the PR. So far it seems to solve more bug than to introduce new one ;)

Aced14 commented:
@gregory38 [Monster Attack: Block dump](https://aced14.github.io/pcsx2-gs-dumps/ma_block_fmv-flicker-bug.7z).

From what I researched, it seems that only the [Linuz ISO CDVD plugin](http://wiki.pcsx2.net/index.php/Setting_up_Windows_version#Linuz_ISO_CDVD) can read/write block dumps. But it's not included in modern PCSX2 builds. I copied it out of a summer 2014 build and used it within this PR's latest AppVeyor build.

tabnk commented:
Finally the rules of rose FMV can be render in hardware mode.

baptiste0602 commented:
This PR introduces half-screen issue with Splinter Cell Chaos Theory : half bottom is black.
Tested with (E) version. I will try later if it's happening in (U).

Should I add it to #1339 ?

refractionpcsx2 commented:
> This PR introduces half-screen issue with Splinter Cell Chaos Theory : half bottom is black.
> Tested with (E) version. I will try later if it's happening in (U).
> 
> Should I add it to #1339 ?

Thanks :) I have added it for you to the "Extra" list as it has a link to here.
